feat(plugins): Add unocss plugin (WIP)

adamgreg opened the pull request:
Adds a plugin that can use UnoCSS to generate CSS at different times:
* During AOT builds (enabled by default)
* During server-side renders (disabled by default)
* During client-side rendering (disabled by default)

UnoCSS transformers are not currently supported.

Rebases and succeeds https://github.com/denoland/fresh/pull/1303

miguelrk committed:
feat(plugins): add unocss plugin (WIP)

miguelrk committed:
fix(plugins/unocss): use presetUno instead of presetWind by default

miguelrk committed:
chore(plugins): add comment to inline reset for unocss

miguelrk committed:
chore(plugins): update comment of inline reset for unocss

miguelrk committed:
docs(plugins): add docs for unocss

miguelrk committed:
fix(plugins): remove unnecessary `?bundle&no-check` from `@unocss/preset-uno` import

miguelrk committed:
docs(plugins): remove redundant text for unocss

miguelrk committed:
fix(plugins): remove defaults from unocss plugin

miguelrk committed:
chore(plugins): add `fixture_unocss_hydrate` test

miguelrk committed:
chore(plugins): bump `@unocss/reset` version

Co-authored-by: Lino Le Van <11367844+lino-levan@users.noreply.github.com>

miguelrk committed:
chore(plugins): bump `@unocss/core` version

Co-authored-by: Lino Le Van <11367844+lino-levan@users.noreply.github.com>

miguelrk committed:
chore(plugins): bump `@unocss/core` and `@unocss/preset-uno` version

Co-authored-by: Lino Le Van <11367844+lino-levan@users.noreply.github.com>

miguelrk committed:
chore(plugins): format via `deno fmt`

miguelrk committed:
feat(plugins): add `antfu.unocss` to recommended vscode extensions

miguelrk committed:
chore(ci): try fixing broken types causing CI to fail

miguelrk committed:
chore(plugins): order imports

miguelrk committed:
chore(plugins): use import map instead of url imports for unocss

miguelrk committed:
feat(plugins): add client runtime script to unocss plugin by default

miguelrk committed:
chore(plugins): unify imports form `@unocss/core`

adamgreg committed:
Re-export UserConfig type from UnoCSS plugin

adamgreg committed:
Import @unocss/preset-uno from full URL in test

adamgreg committed:
Update UnoCSS in plugin to 0.55.1

adamgreg committed:
In the UnoCSS plugin pass the config to runtime

-  Define a Config interface for the plugin, which extends UnoCSS UserConfig with a selfURL property
- Add a `defineConfig()` function for help with types, and make it easier to adapt existing UnoCSS config modules for use with the plugin
- Move the runtime script inline

miguelrk committed:
chore(plugins): format via `deno fmt`

miguelrk committed:
Update plugins/unocss.ts

Use `init()` function of `@unocss/runtime` instead of manually injecting the script tag to avoid fetching it (redundantly) in the client, and instead ship runtime with `plugin-unocss.main.ts`.

Co-authored-by: 木杉 <zhmushan@qq.com>

miguelrk committed:
chore(plugins): update querySelector to style[data-unocss-runtime-layer="default"] in fixture_unocss_hydrate

miguelrk committed:
Update docs/latest/examples/using-unocss.md

Co-authored-by: 木杉 <zhmushan@qq.com>

adamgreg committed:
UnoCSS plugin use uno.config.ts

- Import config from uno.config.ts if no config object is explicitly provided
- Always use uno.config.ts as the import source for the browser runtime config
- This avoids the complexity of selfURL, allowing the plugin to use standard Uno config files

adamgreg committed:
UnoCSS plugin: Add defineConfig() helper function

It is the same as found in the unocss package, but that can not currently be safely imported in Deno (due to Node-specific code in the icons preset).

adamgreg committed:
Make UnoCSS plugin function synchronous

adamgreg committed:
UnoCSS Plugin: Use Preact hook for SSR

adamgreg committed:
UnoCSS plugin: Update the inline reset

Use tailwind-compat.css instead of tailwind.css as the source. This resolves a specificity problem affecting button backgrounds: https://github.com/unocss/unocss/issues/2127

adamgreg committed:
Support AOT build in UnoCSS plugin

- Add separate SSR, CSR & AOT config options
- Bump UnoCSS version to 0.56.5

adamgreg committed:
docs: Update UnoCSS plugin docs

miguelrk commented on plugins/unocss.ts:
```diff
@@ -0,0 +1,188 @@
+
+    async renderAsync(ctx) {
+      // Link to CSS file, if AOT mode is enabled
+      const cssLinks = aot ? [{ url: "/uno.css" }] : [];
```
Line 131 should be updated to `const links = aot ? [{ rel: "stylesheet", href: "/uno.css" }] : [];` to match the new `links` and also the new Link format

adamgreg replied:
Thanks! I'd forgotten about that

miguelrk commented on plugins/unocss.ts:
```diff
@@ -0,0 +1,188 @@
+        await ctx.renderAsync();
+      }
+
+      return { scripts, styles, cssLinks };
```
Rename `cssLinks` to `links` here as well

adamgreg committed:
UnoCSS plugin: fix CSS link format in return of renderAsync()

adamgreg committed:
UnoCSS plugin: Fix typo in stylesheet link

adamgreg committed:
UnoCSS plugin updates

- Use middleware to maintain "AOT" behaviour in dev mode
- Use new configResolved() plugin hook to resolve UnoCSS config file
- Reduce effort during renders in SSR mode

adamgreg committed:
UnoCSS plugin: Check config file exists for CSR mode
